### PR TITLE
Support affine permutations in transforms

### LINKED ISSUE #3410: possible bug in affine transform slicing
## 🐛 Bug

When trying to open a volume with a supplied affine matrix of shape (4, 4), some permutations of axes can result in a crash.  I understand that general affines are not currently supported, but in this case, the data is aligned along the orthogonal axes, just in a permuted order. Is this currently expected to work? 

cc @andy-sweet (I mentioned this in the architecture working group meeting earlier today)

## To Reproduce

Below is a minimal example with an affine that permutes the order of the dimensions in the data (e.g. axis 2 of data corresponds to axis 0 in world coordinates, etc.).

```Python
import numpy as np
import napari
viewer = napari.Viewer()

vol = np.ones((128, 128, 128))

affine = [
    [0, 0, 1, 0],
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 0, 1]
]
viewer.add_image(vol, affine=affine)

napari.run()
```
This fails with the following traceback:

```
napari/napari/layers/base/base.py:745: UserWarning: Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.
  warnings.warn(
Traceback (most recent call last):
  File "affine_crash.py", line 13, in <module>
    viewer.add_image(vol, affine=affine)
  File "napari/napari/components/viewer_model.py", line 753, in add_image
    self.layers.append(layer)
  File "_collections_abc.py", line 962, in append
    self.insert(len(self), value)
  File "napari/napari/components/layerlist.py", line 80, in insert
    super().insert(index, new_layer)
  File "napari/napari/utils/events/containers/_selectable_list.py", line 63, in insert
    super().insert(index, value)
  File "napari/napari/utils/events/containers/_evented_list.py", line 179, in insert
    self.events.inserted(index=index, value=value)
  File "napari/napari/utils/events/event.py", line 575, in __call__
    self._invoke_callback(cb, event)
  File "napari/napari/utils/events/event.py", line 607, in _invoke_callback
    _handle_exception(
  File "napari/napari/utils/events/event.py", line 596, in _invoke_callback
    cb(event)
  File "napari/napari/_qt/qt_viewer.py", line 405, in _on_add_layer_change
    self._add_layer(layer)
  File "napari/napari/_qt/qt_viewer.py", line 415, in _add_layer
    vispy_layer = create_vispy_visual(layer)
  File "napari/napari/_vispy/utils.py", line 62, in create_vispy_visual
    return visual_class(layer)
  File "napari/napari/_vispy/vispy_image_layer.py", line 67, in __init__
    self._on_display_change()
  File "napari/napari/_vispy/vispy_image_layer.py", line 90, in _on_display_change
    self.reset()
  File "napari/napari/_vispy/vispy_image_layer.py", line 187, in reset
    self._reset_base()
  File "napari/napari/_vispy/vispy_base_layer.py", line 165, in _reset_base
    self._on_matrix_change()
  File "napari/napari/_vispy/vispy_base_layer.py", line 153, in _on_matrix_change
    self._master_transform.matrix = affine_matrix
  File "site-packages/vispy/visuals/transforms/linear.py", line 401, in matrix
    self.shader_imap()
  File "site-packages/vispy/visuals/transforms/linear.py", line 389, in shader_imap
    fn['inv_matrix'] = self.inv_matrix  # uniform mat4
  File "site-packages/vispy/visuals/transforms/linear.py", line 407, in inv_matrix
    self._inv_matrix = np.linalg.inv(self.matrix)
  File "<__array_function__ internals>", line 5, in inv
  File "site-packages/numpy/linalg/linalg.py", line 545, in inv
    ainv = _umath_linalg.inv(a, signature=signature, extobj=extobj)
  File "site-packages/numpy/linalg/linalg.py", line 88, in _raise_linalgerror_singular
    raise LinAlgError("Singular matrix")
numpy.linalg.LinAlgError: Singular matrix

```

There is no problem if `affine=np.eye(4)` is used instead.

Aside from the crash, I was surprised to see the warning:
`UserWarning: Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.`

## Expected behavior

As this type of permutation does not involve rotation or shear, I thought it would be supported. Is it instead the case that the matrix can only be diagonal with scale factors in the upper (3x3)?

## Environment

Current main branch of napari on linux.

## Additional context

My current workaround is to permute the data dimensions and make the corresponding changes to the affine before calling `add_image`.

An example real-world image with this type of affine is the following file: 
https://github.com/grlee77/parrec_oblique/blob/master/NIFTI/T1W_sag_0_0_0_SENSE_8_1.nii
which corresponds to a single sagittal slice MRI scan. It's affine is:
```
array([[  0.        ,   0.        ,   5.        ,   2.40481853],
       [ -0.42857143,   0.        ,   0.        , 130.22042847],
       [  0.        ,  -0.42857143,   0.        , 120.        ],
       [  0.        ,   0.        ,   0.        ,   1.        ]])
```

### PR DESCRIPTION
**Pull request by grlee77**
# Description

closes #3410

This PR attempts to support cases where the affine is set using a permutation as in #3410. Prior to this PR, this could cause a linear algebra error as seen in that issue. Fixing the behavior is done here by adding some special handling code for permutations in the `set_slice` method for the Affine transform class and making corresponding updates to Layer methods as needed. From the perspective of the viewer, all attributes of `viewer.dims` reflect the axis order **after** the permutation and most things just continue to work as before. However, there are a few places, such as slicing of the original data array where we need to convert from the permuted order back to the original data order. So far these changes have been fairly small and mostly confined to the base Layer class (no changes have been needed in any of the VisPy or Qt-based classes).

As far as I can tell, things are working as expected in interactive use with the GUI, including the ray intersections for 3D views and the "values" reported in the status below the viewer window (i.e. `get_value` and `get_ray_intersections`). 

Aside from the permutation handling, I also added an `is_diagonal` property that allows optimizing some operations of the `Affine` class for the common case when the `linear_matrix` portion is just a diagonal matrix. That part could also be in a separate smaller PR if preferred.

There is also a bug fix related to 0.5 pixel offset in the ray intersection calculations. I could make a separate PR isolating just that change as well.

### alternative proposal

An alternative to the solution here is to manually transpose the data axes so that the affine becomes diagonal rather than remaining a permutation. This could be in, e.g. `add_image`. That would simplify things in terms of avoiding potential confusion between order of dimensions pre vs. post permutation as data and viewer dimensions would never be permuted relative to one another. There are a couple of drawbacks with that approach, though

1.) it is impractical to transpose the full dataset in cases involving large data, especially data that is larger than memory or stored in the cloud.
2.) by transposing and then updating the affine upon load, the user will likely not realize this was done. if the user then saves the data again, but uses the original affine, the transform information would be incorrect.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

There are quite a few new tests covering the various additions to the `napari.utils.transform` classes. For testing with the ViewerModel and Layers, I have been doing this manually so far, verifying that things like the sliders cover the expected ranges and things work as expected. I need to add some unit tests for this, though, so have marked this as WIP. I also need to test a bit more with the other layer types aside from Image and Labels.

## Final checklist:
- [x My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by grlee77** [comment]
Here is a script I have been using to play around with it (it generates synthetic image and labels dated with a affine matrix involving permutations).
```Python
import napari
import numpy as np
from skimage import data, measure, morphology


affine_type = 'permuted'
# affine_type = 'permuted_no_scale_or_translate'
# affine_type = 'scaled'
# affine_type = 'none'
rgb_image = False

if affine_type.startswith('permuted'):
    if affine_type == 'permuted_no_scale_or_translate':
        # permuted, but with uniform scaling and no translation
        p = np.asarray(
            [[0, 1, 0, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1]]
        )
    else:
        # permuted with non-uniform scale and translation
        p = np.asarray(
            [[0, 0.8, 0, 5.5], [0, 0, 2, -10], [3, 0, 0, 13.5], [0, 0, 0, 1]]
        )
    image_kwargs = dict(affine=p)
    shape = (14, 24, 32)
elif affine_type == 'scaled':
    # scaled, but not permuted or translated
    p = np.eye(4)
    p[:3, :3]  = np.diag([0.8, 2, 3])
    shape = (24, 32, 14)  # 14, 24, 32)
    image_kwargs = dict(affine=p)
else:
    # not permuted, but scaled and translated
    p = np.asarray(
        [[3, 0, 0, 13.5], [0, 0.8, 0, 5.5], [0, 0, 2, -10], [0, 0, 0, 1]]
    )
    image_kwargs = dict(affine=p)
    shape = (14, 24, 32)

# apply_permutation(p, (0, 1), (1, 1))
v = napari.Viewer(ndisplay=2, show=True)

# add a random-valued image
rng = np.random.default_rng()
image_shape = shape + (3,) if rgb_image else shape
image = rng.integers(0, 256, size=image_shape, dtype=np.uint8)
v.add_image(image, **image_kwargs)

# add a labels layer
labels = data.binary_blobs(32, n_dim=3)[:shape[0], :shape[1], :shape[2]]
# erode the binary blobs then call measure.label to get a labels image
labels = morphology.binary_erosion(labels)
labels = measure.label(labels)
v.add_labels(labels, **image_kwargs)

napari.run()
```

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -634,6 +634,53 @@ def _dims_displayed_order(self):
...
+    @property
+    def _data_dims_order(self):
+        """Order of the axes into the raw data array.
+
+        For affines without permutation, it will be the same as _dims_order.
+        """
+        if self._data_to_world.is_permutation:
+            perm = self._data_to_world.perm
+            return [perm[d] for d in self._dims_order]
+        return self._dims_order
```
Comment (jni) [inline napari/layers/base/base.py]: I know some linters remove `else`s in situations like this, but I find it easier to read. In my first reading I'd missed the `return` inside the `if` block and it seemed like you were ignoring the work in that block.

```suggestion
        if self._data_to_world.is_permutation:
            perm = self._data_to_world.perm
            return [perm[d] for d in self._dims_order]
        else:
            return self._dims_order
```

Alternatively:

```python
        if self._data_to_world.is_permutation:
            perm = self._data_to_world.perm
            dims_order = [perm[d] for d in self._dims_order]
        else:
            dims_order = self._dims_order
        return dims_order
```

File: napari/layers/base/base.py [hunk 2]
```diff
@@ -1401,6 +1471,9 @@ def get_ray_intersections(
 
         # create the bounding box in data coordinates
         bounding_box = self._display_bounding_box(dims_displayed)
+        if not world:
+            # need 0 to be the corner of a pixel rather than its center
+            position = tuple(p + 0.5 for p in position)
```
Comment (jni) [inline napari/layers/base/base.py]: Yeah, this should be a separate PR and issue, please. 0 *should* be the center of a pixel according to our coordinate conventions? (ie pixels should be considered samples taken at the given coordinates)

File: napari/utils/transforms/transforms.py [hunk 3]
```diff
@@ -597,6 +648,28 @@ def expand_dims(self, axes: Sequence[int]) -> 'Affine':
...
+        return is_diagonal(self.linear_matrix, tol=1e-8)
+
+    @cached_property
+    def is_permutation(self):
+        return is_permutation(
+            self.linear_matrix, tol=1e-8, exclude_diagonal=False
+        )
+
+    @cached_property
+    def perm(self):
```
Comment (jni) [inline napari/utils/transforms/transforms.py]: Can we rename this "permutation"? I don't think it's worth saving the 7 characters at the cost of readability.

**Review by jni — COMMENTED** [review]
@grlee77 I've only had a chance to have a quick look. I'd definitely prefer for the 0.5 offset bug to be properly explained and fixed in a separate PR. Otherwise this looks ok to me, minus `.perm` -> `.permutation`.

Having said that, I wonder whether this could just be done with functions rather than adding methods/properties to the affine transform. `if (perm := get_permutation(tf)):`, where `get_permutation` returns `None` if the transform is not a permutation, and a permutation otherwise?

File: napari/utils/transforms/transform_utils.py [hunk 4]
```diff
@@ -466,3 +466,86 @@ def is_matrix_triangular(matrix):
...
+        on a data matrix would be equivalent to ``data.transpose(c)``.
+    """
+    n = matrix.shape[0]
+    if tol == 0:
+        r, c = np.nonzero(matrix)
+    else:
+        r, c = np.nonzero(np.abs(matrix) >= tol)
+    # permutation matrix will only have 1 entry per column
+    c = tuple(c)
+    c_check = len(list(set(c))) == n
```
Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
    c_check = len(set(c)) == n
```

File: napari/utils/transforms/transforms.py [hunk 5]
```diff
@@ -405,17 +421,33 @@ def ndim(self) -> int:
...
 
     @scale.setter
     def scale(self, scale):
         """Set the scale of the transform."""
-        rotate, _, shear = decompose_linear_matrix(
-            self.linear_matrix, upper_triangular=self._upper_triangular
-        )
-        self._linear_matrix = compose_linear_matrix(rotate, scale, shear)
+        if self.is_diagonal:
+            for i in range(len(scale)):
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py:484]: You might want to do `scale = scale_to_vector(scale, ndim=self.ndim)` to ensure the scale is padded as in `compose_linear_matrix`.

File: napari/components/_tests/test_layers_list.py [hunk 6]
```diff
@@ -486,18 +486,17 @@ def test_world_extent_mixed_ndim():
...
     layers = LayerList()
 
     layer = Image(
-        np.random.random((15, 15)), affine=[[0, 1, 0], [1, 0, 0], [0, 0, 1]]
+        np.random.random((15, 15)), affine=[[0, 1, 0], [2, 0, 0], [0, 0, 1]]
     )
     layers.append(layer)
-    np.testing.assert_allclose(layer._data_to_world.scale, (1, -1))
-    np.testing.assert_allclose(layers.extent.step, (1, 1))
+    # scale and step reflect the values AFTER the affine has been applied.
```
Comment (andy-sweet) [inline napari/components/_tests/test_layers_list.py:496]: I'm struggling to understand what this means exactly.

In the docstring of `Affine` we describe the transform as being the composition of a few components in a predefined order: `rotate * shear * scale + translate`.

Are we changing this order in general to something like `scale * rotate * shear * permutation + translate`?

File: napari/components/_tests/test_layers_list.py [hunk 7]
```diff
@@ -486,18 +486,17 @@ def test_world_extent_mixed_ndim():
...
-    # calculation
     np.random.seed(0)
     layers = LayerList()
 
     layer = Image(
-        np.random.random((15, 15)), affine=[[0, 1, 0], [1, 0, 0], [0, 0, 1]]
+        np.random.random((15, 15)), affine=[[0, 1, 0], [2, 0, 0], [0, 0, 1]]
     )
     layers.append(layer)
-    np.testing.assert_allclose(layer._data_to_world.scale, (1, -1))
```
Comment (andy-sweet) [inline napari/components/_tests/test_layers_list.py:499]: I'm curious how pure reflections are handled - i.e. what happens for `affine = [[1, 0, 0], [0, -1, 0], [0, 0, 1]]`? Does the reflection still get absorbed into `scale`? Can we have an explicit test for that if it doesn't already exist?

**Review by andy-sweet — COMMENTED** [review]
Thanks for your effort here @grlee77!

The tests are really helpful in understanding some of the changes here, and also help clarify the changes I don't understand yet.

I haven't reviewed all the changes yet because this is still draft and because I have some high level questions that are probably best answered before I continue.

In general, any PR that touches some of the code paths here (i.e. how transforms/layers/slicing interact) always confuses me a little, and unfortunately this one isn't an exception. That could definitely be down to my own ignorance and something I could fix. But I'm curious to know your perspective on how easy you found these changes to make and how happy you are with the solution. If everyone find PRs related to this logic hard to read and write, it suggests ensuring that redesign this logic is prioritized.

**Comment by andy-sweet** [comment]
> Aside from the permutation handling, I also added an `is_diagonal` property that allows optimizing some operations of the `Affine` class for the common case when the `linear_matrix` portion is just a diagonal matrix. That part could also be in a separate smaller PR if preferred.

If this is a pure optimization, I'd strongly suggest separating it into a different PR to make the logic/semantic changes in this PR easier to understand.

**Comment by grlee77** [comment]
Thanks, for the initial comments. I will split a couple of the things into smaller PRs to help make reviewing what is remaining here easier.

> But I'm curious to know your perspective on how easy you found these changes to make and how happy you are with the solution. If everyone find PRs related to this logic hard to read and write, it suggests ensuring that redesign this logic is prioritized.

The changes to the transform classes and utilities were not too difficult. I cannot say the same for the changes to layers/viewer, though. I ended up trying a few different things and even just completely starting over a couple of times! I would say > 15 hours just on that part, despite the relatively few lines of code there. Part of that was me figuring out how things currently operate, but mainly it was not always obvious where I needed to switch to using dimensions ordered as in `ViewerModel.dims` vs. the order in `Layer.data`. 

My first prototype attempt was to make no changes to the transform classes and just try handling as a special case in places where `set_slice` was being called on the transform. This resulted in duplicating various code, so I wanted to refactor it to live on the transform class itself.

Just changing the transform's `set_slice` was not sufficient as things like layer's `extent.data` would still have axes in a different order than `extent.world`! In the current PR this is handled by checking if the `Layer._data_to_world` transform is a permutation and making `extent.data` have the same axis order as `extent.world`. This helps make sure that all downstream things relying on these continue to work as before and I didn't need to modify anything in terms of Qt or VisPy code. Specifically, logic like which axes the sliders correspond to and how the camera center is calculated, etc. all stay the same.

Currently, with the exception of `Layer._data` itself, pretty much everything is in the world coordinate axis order. The issue was that things like `_dims_displayed` or `_dims_not_displayed` were sometimes used as the dimensions for world space, but were also used when setting slices in `Layer._data` itself. They key thing here was figuring out when the original data dimension order needed to be used instead (mainly in the indices used for slicing within `ImageBase._set_view_slice`).

Another quite tricky part was that for `get_ray_intersections` when `world=False` the ranges correspond to those of the data (e.g. without the `scale` applied), but the axis order is still the same as in world coordinate space! This was because these functions interact **not** with the original raw `Layer._data` as I had assumed, but instead with `ImageBase._slice.image.raw` which has [already been transposed](https://github.com/napari/napari/blob/0ff842d11c6e67c77373ffda011125ee9d7bc98b/napari/layers/image/image.py#L793) to have axes matching the displayed order!

File: napari/utils/transforms/transforms.py (hunk 5 above)
Comment (grlee77) [reply napari/utils/transforms/transforms.py:484]: Good catch, done in [8825912](https://github.com/napari/napari/pull/4303/commits/8825912f1701cb63d62e60ba9b803bd940c34888)

File: napari/components/_tests/test_layers_list.py (hunk 6 above)
Comment (grlee77) [reply napari/components/_tests/test_layers_list.py:496]: I think we still need to decide what we want this to look like. Right now there is a `data2physical` `CompositeAffine` transform which is `rotate * shear * scale + translate` like you say. Then there is also a separate `physical2world` transform that is a general `Affine` (although in practice only a diagonal physical2world works correctly prior to this PR). So far, I was testing with setting that separate `physical2world` matrix to be a permutation rather than the default of an identity matrix.

I think it would make sense that `data2physical` itself could involve a permutation, but haven't tried to refactor `CompositeAffine` to add a separate permutation property there.

File: napari/components/_tests/test_layers_list.py (hunk 7 above)
Comment (grlee77) [reply napari/components/_tests/test_layers_list.py:499]: yes, scale in this case would be [1, -1, 1]. I can add a test case with negative scales.

```Python
import numpy as np
from napari.utils.transforms import Affine

linear_matrix = np.asarray([[1, 0, 0], [0, -1, 0], [0, 0, 1]])
A = Affine(linear_matrix=linear_matrix)
print(A.scale)
```

This ends up being the same as the scale given by:
```Python
from napari.utils.transforms.transform_utils import decompose_linear_matrix

rotate, scale, translate = decompose_linear_matrix(linear_matrix)
```

**Comment by grlee77** [comment]
> If this is a pure optimization, I'd strongly suggest separating it into a different PR to make the logic/semantic changes in this PR easier to understand.

I can split the `is_diagonal` optimization into a separate small PR

File: napari/utils/transforms/transforms.py (hunk 3 above)
Comment (grlee77) [reply napari/utils/transforms/transforms.py]: Yes, done!

File: napari/components/_tests/test_layers_list.py (hunk 6 above)
Comment (grlee77) [reply napari/components/_tests/test_layers_list.py:496]: > but haven't tried to refactor CompositeAffine to add a separate permutation property there.

To clarify, there is the read-only permutation property as introduced in this PR that returns the axis order, but there is no permutation argument to the constructor. To assign a different permutation to the Affine after it has already been created, one would have to either assign `linear_matrix` or one could assign both `rotation` and `scale` as determined via `decompose_linear_matrix`

**Comment by grlee77** [comment]
Now that the #4315 and #4370 have been merged, I fixed the conflicts and this should be ready for review.

I added a couple new test cases covering Image layer extents and viewer Dims properties.

File: napari/utils/transforms/transforms.py [hunk 8]
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
...
         Can be overriden in subclasses to enable performance optimizations
         that are specific to this case.
         """
         return False
 
+    @property
+    def _is_permutation(self) -> bool:
+        """Indicate a transform corresponding to a permutation of dimensions.
+
+        Can be overriden in subclasses to enable performance optimizations
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
        Can be overridden in subclasses to enable performance optimizations
```

File: napari/utils/transforms/transforms.py [hunk 9]
```diff
@@ -135,11 +157,15 @@ def inverse(self) -> 'TransformChain':
         return TransformChain([tf.inverse for tf in self[::-1]])
 
     @property
-    def _is_diagonal(self):
+    def _is_diagonal(self) -> bool:
         if all(getattr(tf, '_is_diagonal', False) for tf in self):
             return True
         return getattr(self.simplified, '_is_diagonal', False)
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
        return all(tf._is_diagonal for tf in self)
```

As `_is_diagonal` is a property defined on the base class that has a default implementation, can we simplify this? Or maybe `return self.simplified._is_diagonal`?

File: napari/utils/transforms/transforms.py [hunk 10]
```diff
@@ -135,11 +157,15 @@ def inverse(self) -> 'TransformChain':
...
 
+    @property
+    def _is_permutation(self) -> bool:
+        return getattr(self.simplified, '_is_permutation', False)
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
        return self.simplified._is_permutation
```

As `_is_permutation` is defined on the base class, can we simplify?

File: napari/utils/transforms/transforms.py [hunk 11]
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
         )
 
     @property
-    def _is_diagonal(self):
+    def _is_diagonal(self) -> bool:
         """Indicate when a transform does not mix or permute dimensions.
 
         Can be overriden in subclasses to enable performance optimizations
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
        Can be overridden in subclasses to enable performance optimizations
```

File: napari/utils/transforms/transforms.py [hunk 12]
```diff
@@ -445,6 +485,12 @@ def scale(self, scale):
             scale = scale_to_vector(scale, ndim=self.ndim)
             for i in range(len(scale)):
                 self._linear_matrix[i, i] = scale[i]
+        elif self._is_permutation:
+            # assumes scale[i] is the scaling on axis i AFTER the transform
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py:487]: After what transform? After the permutation?

File: napari/utils/transforms/transform_utils.py [hunk 13]
```diff
@@ -490,3 +490,66 @@ def is_diagonal(matrix, tol=1e-8):
...
+    """
+    n = matrix.shape[0]
+    if tol == 0:
+        r, c = np.nonzero(matrix)
+    else:
+        r, c = np.nonzero(np.abs(matrix) >= tol)
+    # permutation matrix will only have 1 entry per column
+    c = tuple(c)
+    c_check = len(set(c)) == n
+    if not c_check:
```
Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
    if len(set(c)) != n:
```

Optional: simplify or at least consider a more meaningful name than `c_check`.

File: napari/utils/transforms/transforms.py [hunk 14]
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
...
+        """Indicate a transform corresponding to a permutation of dimensions.
+
+        Can be overriden in subclasses to enable performance optimizations
+        that are specific to this case.
+        """
+        return False
+
+    @property
+    def _permutation(self) -> Optional[Tuple[int]]:
+        """Indicate when a transform does not mix or permute dimensions.
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py:120]: The summary line seems to be copy/pasted from `_is_permutation`? Consider just removing the summary and keeping the rest.

File: napari/components/_tests/test_layers_list.py (hunk 7 above)
Comment (andy-sweet) [reply napari/components/_tests/test_layers_list.py:499]: A short extra test seems useful to me.

File: napari/layers/base/base.py [hunk 15]
```diff
@@ -754,7 +804,11 @@ def _slice_indices(self):
                     category=UserWarning,
                 )
 
-        slice_inv_transform = inv_transform.set_slice(self._dims_not_displayed)
+        slice_inv_transform = inv_transform.set_slice(
+            self._data_dims_not_displayed
+        )
+
+        # _dims_point is already permuted so don't use _permuted_dims_not_displayed
```
Comment (andy-sweet) [inline napari/layers/base/base.py:811]: `_permuted_dims_not_displayed` does not exist, but `_data_dims_not_displayed` does. Either update the comment. Or maybe better consider renaming the new properties on `Layer` to start with `_permuted` instead of `_data`?

File: napari/utils/transforms/transforms.py (hunk 12 above)
Comment (andy-sweet) [reply napari/utils/transforms/transforms.py:487]: Does this effectively mean that when `_is_permutation` is true then we can always have `linear_matrix = scale * permutation`?

I.e. the `rotate` and `shear` components will always be identity?

File: napari/utils/transforms/transforms.py [hunk 16]
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
...
         """Indicate when a transform does not mix or permute dimensions.
 
         Can be overriden in subclasses to enable performance optimizations
         that are specific to this case.
         """
         return False
 
+    @property
+    def _is_permutation(self) -> bool:
+        """Indicate a transform corresponding to a permutation of dimensions.
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py:111]: I need a more clear/explicit description here. I think this returns true if the transforms is a "pure" permutation (that also allows scaling afterwards), but that does not allow things like rotation and shear that mix dimensions - is that right?

File: napari/layers/image/image.py [hunk 17]
```diff
@@ -729,11 +771,11 @@ def _set_view_slice(self):
...
             )
             indices[not_disp] = downsampled_indices
-
-            thumbnail_source = self.data[self._thumbnail_level][tuple(indices)]
+            thumb_indices = self._permute_indices(indices)
+            thumbnail_source = self.data[self._thumbnail_level][thumb_indices]
         else:
             self._transforms['tile2data'].scale = np.ones(self.ndim)
-            image_indices = self._slice_indices
+            image_indices = self._permute_indices(self._slice_indices)
```
Comment (andy-sweet) [inline napari/layers/image/image.py:782]: If I understand correctly, `_slice_indices` has already absorbed any permutation, so is `_permute_indices` inverting that so we can index using the non-permuted dimensions?

File: napari/components/_tests/test_layers_list.py (hunk 6 above)
Comment (andy-sweet) [reply napari/components/_tests/test_layers_list.py:496]: Thanks for the comments. I think I understand things a little better now, though that may have generated more clarifying questions elsewhere in this PR.

File: napari/layers/image/image.py (hunk 17 above)
Comment (andy-sweet) [reply napari/layers/image/image.py:782]: Does this only need to happen for images? Have you tested with other layer types (e.g. points)?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Thanks for splitting up this PR and getting the simpler parts in first. I understand the changes here better now, but am still a little confused/scared by this PR.

File: napari/components/_tests/test_layers_list.py (hunk 7 above)
Comment (grlee77) [reply napari/components/_tests/test_layers_list.py:499]: added this test in [306c14a](https://github.com/napari/napari/pull/4303/commits/306c14a6aaa9268eadda07fdd7a563bf8e5e84a7)

File: napari/layers/base/base.py (hunk 15 above)
Comment (grlee77) [reply napari/layers/base/base.py:811]: conclusion during the last architecture working group meeting was not to do this renaming

File: napari/layers/image/image.py (hunk 17 above)
Comment (grlee77) [reply napari/layers/image/image.py:782]: I think there is a bug here in the multiscale case (I'm pretty sure it should be `image = self.data[level][image_indices]` rather than `image = self.data[level][indices]`). I will add test coverage for permutations when the data is multiscale.

**Comment by grlee77** [comment]
**TODO**
- [x] implement for Points layers
- [ ] implement for Shapes layers
- [x] implement for Surface layers
- [x] implement for Tracks layers
- [x] implement for Vectors layers

I did manual testing via the GUI for the layers implemented above, but would ideally add additional tests as well:
- [ ] add tests for Points layers
- [ ] add tests for Shapes layers
- [ ] add tests for Surface layers
- [ ] add tests for Tracks layers
- [ ] add tests for Vectors layers

**Comment by maweigert** [comment]
Hi, 

Just a random comment (as I was wondering what kind of affine matrix was permissible in `add_image` too), but I think one could obtain the permutation directly from any invertible matrix via [LUP factorization](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu.html). I.e. for an invertible matrix `A` (being the 3x3 spatial part of the affine matrix) this would give you the decomposition `A=P*L*U` with `P` being a pure permutation and `B=L*U` the (invertible) spatial part. An additional benefit is that `B` [should have invertible subminors](https://en.wikipedia.org/wiki/LU_decomposition#Square_matrices) and thus `B.set_slice` should always return a non-singular matrix (this way, any invertible matrix would be permissible in `add_image`).

**Comment by Carreau** [comment]
This PR have not been active for quite some time. 

In order to keep the list of PRs short and easier to work with, I am going to tag this PR as [stale](https://github.com/napari/napari/pulls?q=is%3Apr+label%3Astale+is%3Aclosed) and close it.

Feel free to reopen and remove the "stale" label when work on this PR resumes. 
If you do not have the permissions to do so, feel free to request help from one of the maintainer to reopen and remove the `stale` label .
